Content.Server/Research/Systems/ResearchSystem.cs

```diff
@@ -78,19 +79,32 @@ public string[] GetServerNames()
         /// Gets the ids of all the servers
         /// </summary>
         /// <returns></returns>
-        public int[] GetServerIds()
+        public int[] GetServerIds(EntityUid client)
         {
-            var allServers = EntityQuery<ResearchServerComponent>(true).ToArray();
+            var allServers = GetServers(client).ToArray();
             var list = new int[allServers.Length];
 
             for (var i = 0; i < allServers.Length; i++)
             {
-                list[i] = allServers[i].Id;
+                list[i] = allServers[i].Comp.Id;
             }
 
             return list;
         }
 
+        public IEnumerable<Entity<ResearchServerComponent>> GetServers(EntityUid client)
+        {
+            var clientXform = Transform(client);
+            var query = AllEntityQuery<ResearchServerComponent, TransformComponent>();
+            while (query.MoveNext(out var uid, out var comp, out var xform))
+            {
+                if (xform.GridUid != clientXform.GridUid)
+                    continue;
+
+                yield return (uid, comp);
+            }
+        }
+
```

Comment:
Don't we have GetEntitiesInGrid for dis.

Reply:
:mistake: